### PR TITLE
feat: Adding matching WU Context Menu functionality from the ECL IDE

### LINKED ISSUE #363: Match context menu items on WorkUnits from the Windows ECL IDE
Try to match as close as possible the context menu found on the ECL IDE for windows in the VSCode ECL Language Extension.
ECLIDEWorkunitsContextMenu

### PR DESCRIPTION
**Pull request by dehilsterlexis**
Fixes #363

### CODE REVIEW

**Comment by dehilsterlexis** [comment]
@GordonSmith I need to put a limit check still on the CSV files. I will let you know when that is done.

**Comment by dehilsterlexis** [comment]
@GordonSmith I fixed my sequential problem so this is ready for full review.

**Comment by dehilsterlexis** [comment]
@GordonSmith for review

File: package.json [hunk 1]
```diff
@@ -44,12 +44,14 @@
...
-    "@vscode/webview-ui-toolkit": "1.2.2"
+    "@vscode/webview-ui-toolkit": "1.2.2",
+    "csv": "^6.3.5",
+    "exceljs": "^4.4.0"
   },
   "devDependencies": {
     "@fluentui/react": "8.111.2",
     "@hpcc-js/common": "2.71.11",
-    "@hpcc-js/comms": "2.84.2",
+    "@hpcc-js/comms": "^2.85.0",
```
Comment (GordonSmith) [inline package.json:54]: Remove "^"

File: package.json [hunk 2]
```diff
@@ -87,6 +89,7 @@
     "copy-to-clipboard": "3.3.3",
     "copyfiles": "2.4.1",
     "css-loader": "6.8.1",
+    "csv-writer": "^1.6.0",
```
Comment (GordonSmith) [inline package.json:92]: Remove "^"

File: package.json [hunk 3]
```diff
@@ -356,6 +359,36 @@
...
       },
+      {
+        "command": "hpccPlatform.browseECLWatch",
+        "category": "ECL",
+        "title": "%Browse ECL Watch%"
+      },
+      {
+        "command": "hpccPlatform.openResults",
+        "category": "ECL",
+        "title": "%Open Results%"
```
Comment (GordonSmith) [inline package.json:370]: Shoud this be "Browse Results"?  IOW does it open in an external browser, or does it activate the results view in vscode?

File: package.json [hunk 4]
```diff
@@ -1093,19 +1142,7 @@
...
-          "type": "string",
-          "scope": "resource",
-          "enum": [
-            "internal",
-            "external",
-            "disabled"
-          ],
-          "default": "internal",
-          "description": "%Show results after submitting a workunit%"
+          "description": "%Automatically open Workunits on creation%"
```
Comment (GordonSmith) [inline package.json:1145]: Should be "Automatically open Results on creation"?

File: src/ecl/saveData.ts [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+
+export let saveData: SaveData;
+export class SaveData {
+    private _wu: Workunit;
+    private _saveType: SaveType;
+    private _type: SaveFileType;
+    private _filePath: string;
+    private _resultCount: number;
+    private _numResults: number;
+    private _workbook;
```
Comment (GordonSmith) [inline src/ecl/saveData.ts:19]: Missing type

File: src/ecl/saveData.ts [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+    private _sheet: ExcelJS.WorksheetView;
+    private _csvWriter;
```
Comment (GordonSmith) [inline src/ecl/saveData.ts:21]: Missing type

File: src/ecl/saveData.ts [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+    private _createCsvWriter;
```
Comment (GordonSmith) [inline src/ecl/saveData.ts:22]: Not needed

File: src/ecl/saveData.ts [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+    private _dataRows;
```
Comment (GordonSmith) [inline src/ecl/saveData.ts:23]: Shouldn't be needed?

File: src/ecl/saveData.ts [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+        this._wu = wu;
+    }
+
+    async askFileSave(resultName: string) {
+        this._type = SaveFileType.UNKNOWN;
+        const filePath = this._saveType == SaveType.RESULT ? this.filePathResult(this._wu.Wuid, resultName) : this._wu.Wuid;
+        const options: vscode.OpenDialogOptions = {
+            defaultUri: vscode.Uri.file(filePath),
+            canSelectMany: false,
+            openLabel: "Open",
```
Comment (GordonSmith) [inline src/ecl/saveData.ts:37]: Translation

File: src/ecl/saveData.ts [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+            filters: {
+                "CSV Files": ["csv"],
+                "Excel Workbooks": ["xlsx"]
```
Comment (GordonSmith) [inline src/ecl/saveData.ts:40]: Translation

File: src/ecl/saveData.ts [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+        const totalRows = result.Total > maxRows ? maxRows : result.Total;
+        const numPieces = Math.floor(totalRows / pieceSize);
+        const pieces = numPieces == 0 ? 1 : numPieces;
+        let i = 0;
+        let number = 0;
+        let displayNum = 0;
+
+        if (this._type == SaveFileType.CSV) {
+            this._createCsvWriter = createArrayCsvWriter;
+            eclStatusBar.showClientTools(`${result.Name} saving`, "CSV File");
```
Comment (GordonSmith) [inline src/ecl/saveData.ts:113]: Translation

File: src/ecl/saveData.ts [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+            this._dataRows = [];            
+        } else {
+            this._sheet = this._workbook.addWorksheet(result.Name);
+        }
+
+        for (let s = 0; i < pieces; s += pieceSize, i++) {
+            number = totalRows < pieceSize ? totalRows : pieceSize;
+            displayNum = s + pieceSize > totalRows ? totalRows : s + pieceSize;
+            eclStatusBar.showClientTools(`${result.Name} (${displayNum} of ${totalRows})...`, `Saving data to ${this._filePath}`);
```
Comment (GordonSmith) [inline src/ecl/saveData.ts:122]: Translation

File: src/ecl/saveData.ts [hunk 13]
```diff
@@ -0,0 +1,247 @@
...
+            await this.fetchRows(result, s, number);
+        }
+
+        if (this._type == SaveFileType.CSV) {
+            this._csvWriter
+                .writeRecords(this._dataRows)
+                .then(() => console.log("CSV file successfully written"))
```
Comment (GordonSmith) [inline src/ecl/saveData.ts:129]: Translation

**Review by GordonSmith — CHANGES_REQUESTED** [review]

File: package.json (hunk 4 above)
Comment (dehilsterlexis) [reply package.json:1145]: Agreed. That would be clearer although that was not part of this task.

File: src/ecl/saveData.ts (hunk 5 above)
Comment (dehilsterlexis) [reply src/ecl/saveData.ts:19]: That would be ExcelJS.Workbook but that is not an assignable type.

File: package.json (hunk 3 above)
Comment (GordonSmith) [reply package.json:370]: Can you open a new ticket to "Browse WU Details" for WUs

File: src/ecl/saveData.ts [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+            eclStatusBar.showClientTools(`${result.Name} (${displayNum} of ${totalRows})...`, `${localize("Saving data to")} ${this._filePath}`);
+            await this.fetchRows(result, s, number);
+        }
+
+        if (this._type == SaveFileType.CSV) {
+            this._csvWriter
+                .writeRecords(this._dataRows)
+                .then(() => console.log(localize("CSV file successfully written")))
+                .catch((err) => console.error(localize("Error writing CSV file:"), err));
+            eclStatusBar.showClientTools(`${result.Name} saved`, "CSV File");
```
Comment (GordonSmith) [inline src/ecl/saveData.ts:134]: Missing localize for "saved"

File: src/ecl/saveData.ts [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+        }
+        return userLimit;
+    }
+
+    async promptMaxRows(maxRowsFound: number): Promise<number> {
+        if (maxRowsFound <= this._maxRows)
+            return maxRowsFound;
+        const rows = await vscode.window.showInputBox({
+            title: localize("Max number of rows to save"),
+            prompt: `${maxRowsFound} ${localize("rows found. Please choose")} ${this._maxRows} ${localize("or less")}`,
```
Comment (GordonSmith) [inline src/ecl/saveData.ts:227]: For future ref:  Another option is to have a single string "MMM rows found. Please choose NNN or less" and swap out the MMM and NNN with numbers (can be easier for the translators).

**Review by GordonSmith — CHANGES_REQUESTED** [review]

File: src/ecl/saveData.ts (hunk 14 above)
Comment (dehilsterlexis) [reply src/ecl/saveData.ts:134]: Done.

**Comment by dehilsterlexis** [comment]
@GordonSmith for review

File: package.json (hunk 3 above)
Comment (GordonSmith) [reply package.json:370]: Issue for vscode-ecl

**Review by GordonSmith — APPROVED** [review]

**Pull request merged**